### DISCUSSION TITLE
imported package deploys as if it has no module parent.

### DISCUSSION THREAD

**Discussion by joshyiznitsky** [serverless/serverless#12326] [Q&A] [answered]
Hello, I'm working on adding pdf parsing functionality to a project that uses the Serverless Framework. The package I am using to add this functionality is "pdf-parse" and I am importing it into my chunk.ts file by using the statement "import PdfParse from 'pdf-parse' ". My project is written in typescript and deploys on AWS. during deployment, the code is converted to js and then deployed to a Lambda function. When the Lambda function runs I keep getting an ENOENT error. I found that the file it is looking for is one of the test .pdf files in the pdf-parse package, and that should only be triggered if the module has no parent. In this case, I am importing the pdf-parse module, so this code should not be running. I am not sure why this is happening and would appreciate any insight into why the package isn't importing correctly. When it runs correctly it should parse the text from the pdf document when it is called.

The error I am getting in CloudWatch after the resulting Lambda is triggered is this:

{
    "errorType": "Error",
    "errorMessage": "ENOENT: no such file or directory, open './test/data/05-versions-space.pdf'",
    "code": "ENOENT",
    "errno": -2,
    "syscall": "open",
    "path": "./test/data/05-versions-space.pdf",
    "stack": [
        "Error: ENOENT: no such file or directory, open './test/data/05-versions-space.pdf'",
        "    at Object.openSync (node:fs:603:3)",
        "    at Object.readFileSync (node:fs:471:35)",
        "    at node_modules/pdf-parse/index.js (/var/task/src/handlers/chunk.js:282056:27)",
        "    at __require (/var/task/src/handlers/chunk.js:18:50)",
        "    at Object.<anonymous> (/var/task/src/handlers/chunk.js:282126:16)",
        "    at Module._compile (node:internal/modules/cjs/loader:1256:14)",
        "    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)",
        "    at Module.load (node:internal/modules/cjs/loader:1119:32)",
        "    at Module._load (node:internal/modules/cjs/loader:960:12)",
        "    at Module.require (node:internal/modules/cjs/loader:1143:19)"
    ]
}

The base code for the project I am modifying can be found at this link:
https://github.com/mikaelvesavuori/bedrock-rag-demo

I have made minimal changes to enable pdf parsing and have attached my modified chunk.ts file to this post.
[chunk.ts.zip](https://github.com/serverless/serverless/files/13956083/chunk.ts.zip)

Thanks in advance for your help

**Comment by dominicfarr** [comment]
Did you find a solution to this? It seems to fire on import.

`import Pdf from 'pdf-parse';`

**Reply by connect-abdulbasit** [reply]
I resolved the issue by adding the following configuration to next.config.js:
```
module.exports = {
experimental: {
serverComponentsExternalPackages: ['pdf-parse'],
},
};
```
This prevented Next.js from incorrectly trying to access a local file, and everything worked fine after that.

**Comment by joshyiznitsky** [comment]
I haven’t, but I agree with you. It seems as though “isDebugMode” is
getting set to true, triggering the tests on import but I haven’t figured
out how to prevent it.

**Comment by ReasonDuan** [comment]
I also encountered the same problem

**Comment by chegger** [comment]
Did you ever find a solution to this?

**Comment by somalil** [comment]
Just change the isDebugMode variable to false in node_modules/pdf-parse/index.js

**Comment by darkostanimirovic** [comment]
`isDebugMode` is false in my dist but it still fires the test on import...

**Comment by MusiCode1** [comment]
I understood the problem.
The `parent` property of `module` exists only in CommonJS and not in ES Module.

That's why you must use CommonJS.

**Comment by ulimonte05** [comment]
If you delete the test part of code that gives the error (in the pdf-parser node-modules), would fix.

**Comment by diegocasmo** [comment]
`import * as parser from "pdf-parse/lib/pdf-parse.js";` works for me

**Comment by thesrs02** [comment]
`import * as pdfParse from "pdf-parse/lib/pdf-parse.js";`
`pdfParse.default(data)`

if still getting ` pdfParse is not a function` after @diegocasmo  suggestion.

**Reply by pierce-erezhenderson** [reply]
Thank you!

**Reply by Avikalp-Karrahe** [reply]
> Thank you!

did you find a solution?

**Comment by MdTehseenKhan** [comment]
I faced the same problem while working with the `pdf-parse` npm package on a Next.js project to extract text from PDF documents. Here's how I resolved this error on my end. I went to the `pdf-parse/lib/pdf-parse.js` file and separated the code for my use case.
I removed the `pdf-parse` package and instead used a pre-built version of the `pdf.js` library. It provides pre-built versions through NPM and Bower under the name `pdfjs-dist`.

```bash
npm install pdfjs-dist
```

```typescript
// @/utils/pdf-parse.ts

import * as PDFJS from 'pdfjs-dist';

PDFJS.GlobalWorkerOptions.workerSrc = 'pdfjs-dist/build/pdf.worker.mjs';

export async function pdfParse(pdfBuffer: ArrayBuffer): Promise<string> {
  const loadingTask = PDFJS.getDocument({ data: pdfBuffer });
  const pdfDocument = await loadingTask.promise;

  let fullText = '';

  for (let pageNum = 1; pageNum <= pdfDocument.numPages; pageNum++) {
    const page = await pdfDocument.getPage(pageNum);
    const pageText = await extractTextFromPage(page);
    fullText += pageText + '\n\n';
  }

  return fullText.trim();
}

async function extractTextFromPage(page: PDFJS.PDFPageProxy): Promise<string> {
  const textContent = await page.getTextContent();
  return textContent.items.map((item: any) => item.str).join(' ');
}
```


For Next.js:
```typescript
// next.config.js

module.exports = {
  webpack: (config) => {
    config.resolve.alias.canvas = false;
    return config;
  },
};
```

**Comment by connect-abdulbasit** [comment, marked as answer]
I got the same error working with pdf-parse in next  js which i fixed by updating my next.config.ts:

```
module.exports = {
experimental: {
serverComponentsExternalPackages: ['pdf-parse'],
},
};`
```
This prevented Next.js from incorrectly trying to access a local file, and everything worked fine after that.

**Reply by MdTehseenKhan** [reply]
Yeah, it is a better solution.

**Reply by iaurg** [reply]
This really solved, thanks!

**Comment by Granato25** [comment]
I’ve struggled with local PDF parsing in serverless/Node environments too — inconsistent behavior, missing test assets paths, and deployment errors can make it brittle.
What helped me was using a very minimal client that offloads parsing and just consumes JSON results instead of doing heavy work locally.
Sharing here: https://github.com/Granato25/fast-pdf-parser-node
